pjlib/src/pj/ssl_sock_ossl.c

```diff
@@ -455,12 +455,97 @@ static int openssl_init_count;
 /* OpenSSL application data index */
 static int sslsock_idx;
 
+#if defined(PJ_SSL_SOCK_USE_THREAD_LOCK) && PJ_SSL_SOCK_USE_THREAD_LOCK != 0 \
+    && OPENSSL_VERSION_NUMBER < 0x10100000
+
+/* OpenSSL thread lock initialization counter. */
+static int openssl_lock_init_count;
+
+/* OpenSSL locking list. */
+pj_lock_t **ossl_locks;
+
+/* OpenSSL number locks. */
+unsigned ossl_num_locks;
+
+#if     OPENSSL_VERSION_NUMBER >= 0x10000000
+static void ossl_set_thread_id(CRYPTO_THREADID *id)
+{
+    CRYPTO_THREADID_set_numeric(id,
+                     (unsigned long)pj_thread_get_os_handle(pj_thread_this()));
+}
+
+#else
+
+static unsigned long ossl_thread_id(void)
+{
+    return ((unsigned long)pj_thread_get_os_handle(pj_thread_this()));
+}
+#endif
+
+static void ossl_lock(int mode, int id, const char *file, int line)
+{
+    PJ_UNUSED_ARG(file);
+    PJ_UNUSED_ARG(line);
+    if (mode & CRYPTO_LOCK) {
+        if (ossl_locks[id]) {
+            //PJ_LOG(6, (THIS_FILE, "Lock File (%s) Line(%d)", file, line));
+            pj_lock_acquire(ossl_locks[id]);
+        }
+    } else {
+        if (ossl_locks[id]) {
+            //PJ_LOG(6, (THIS_FILE, "Unlock File (%s) Line(%d)", file, line));
+            pj_lock_release(ossl_locks[id]);
+        }
+    }
+}
+
+static pj_status_t init_ossl_lock(pj_pool_t* pool) {    
+    if (openssl_lock_init_count == 0) {        
+        ossl_num_locks = CRYPTO_num_locks();
+        ossl_locks = (pj_lock_t **)pj_pool_calloc(pool,
+                                                  ossl_num_locks,
+                                                  sizeof(pj_lock_t*));
+
+        if (ossl_locks) {
+            unsigned i = 0;
+            for (; i < ossl_num_locks; ++i) {
+                pj_lock_create_simple_mutex(pool, "ossl_lock%p",
+                                            &ossl_locks[i]);
+            }
+#if     OPENSSL_VERSION_NUMBER >= 0x10000000
+            CRYPTO_THREADID_set_callback(ossl_set_thread_id);
+#else
+            CRYPTO_set_id_callback(ossl_thread_id);
+#endif
+
+            CRYPTO_set_locking_callback(ossl_lock);
+        } else {
+            return PJ_ENOMEM;
+        }
+    }
+    openssl_lock_init_count++;
+    return PJ_SUCCESS;
+}
+
+#endif
 
 /* Initialize OpenSSL */
-static pj_status_t init_openssl(void)
+static pj_status_t init_openssl(pj_ssl_sock_t *ssock)
```

Comment:
- init_openssl() context is supposed to be for global initialization, so I don't think the SSL socket param is appropriate here.
- is it safe to use SSL socket pool for allocating global var? i.e: if SSL socket is destroyed, won't the global var memory slot be released too?
  an idea, perhaps just use OpenSSL API to allocate memory (can't remember the func name)? and don't forget to release it later.

Reply:
> * init_openssl() context is supposed to be for global initialization, so I don't think the SSL socket param is appropriate here.
There are 2 places called init_openssl():
* ```ssl_create()``` which accepting socket param
* ```ssl_ciphers_populate()```
> * is it safe to use SSL socket pool for allocating global var? i.e: if SSL socket is destroyed, won't the global var memory slot be released too?
If we intend to use pj_lock_t, then pj_pool_t is required to create them. That is why we use ```openssl_lock_init_count``` to count the number of times the callback init is called. We can only release the callback after the counter set to 0. 
>   an idea, perhaps just use OpenSSL API to allocate memory (can't remember the func name)? and don't forget to release it later.
If we intend to use pj_lock_t for locking, then pj_pool_t is required.

Reply:
> > * init_openssl() context is supposed to be for global initialization, so I don't think the SSL socket param is appropriate here.
> >   There are 2 places called init_openssl():
> 
> * `ssl_create()` which accepting socket param
> * `ssl_ciphers_populate()`
> 

ideally it is PJSIP endpoint init that calls init_openssl(), but it is not necessary for now, so inserting it in ssl_create() is sufficient.

so, still init_openssl() is for global context, not SSL socket specific, don't put the SSL socket as param.

> > * is it safe to use SSL socket pool for allocating global var? i.e: if SSL socket is destroyed, won't the global var memory slot be released too?
>
> If we intend to use pj_lock_t, then pj_pool_t is required to create them. That is why we use `openssl_lock_init_count` to count the number of times the callback init is called. We can only release the callback after the counter set to 0.

if the SSL socket is destroyed, its pool will be released too, right? then what will happen with the pj_lock_t (that may still be used by other any SSL sockets)? that's why it is not safe.

>   an idea, perhaps just use OpenSSL API to allocate memory (can't remember the func name)? and don't forget to release it later.
>  >  If we intend to use pj_lock_t for locking, then pj_pool_t is required.

ya that is okay, but don't use 'short lived' pool such as SSL socket's.

Reply:
> > > * init_openssl() context is supposed to be for global initialization, so I don't think the SSL socket param is appropriate here.
> > >   There are 2 places called init_openssl():
> > 
> > 
> > 
> > * `ssl_create()` which accepting socket param
> > * `ssl_ciphers_populate()`
> 
> ideally it is PJSIP endpoint init that calls init_openssl(), but it is not necessary for now, so inserting it in ssl_create() is sufficient.
> 
> so, still init_openssl() is for global context, not SSL socket specific, don't put the SSL socket as param.
> 
okay
> > > * is it safe to use SSL socket pool for allocating global var? i.e: if SSL socket is destroyed, won't the global var memory slot be released too?
> > 
> > 
> > If we intend to use pj_lock_t, then pj_pool_t is required to create them. That is why we use `openssl_lock_init_count` to count the number of times the callback init is called. We can only release the callback after the counter set to 0.
> 
> if the SSL socket is destroyed, its pool will be released too, right? then what will happen with the pj_lock_t (that may still be used by other any SSL sockets)? that's why it is not safe.
> 
okay
> > an idea, perhaps just use OpenSSL API to allocate memory (can't remember the func name)? and don't forget to release it later.
> > > If we intend to use pj_lock_t for locking, then pj_pool_t is required.
> 
> ya that is okay, but don't use 'short lived' pool such as SSL socket's.

will create a new pool then.